src/ucp/core/ucp_listener.c

```diff
@@ -0,0 +1,156 @@
+     * the main thread
+     */
+    if (listener->cb != NULL) {
+        accept = ucs_malloc(sizeof(*accept), "ucp_listener_accept");
+        if (accept == NULL) {
+            ucs_error("failed to allocate listener accept context");
+            status = UCS_ERR_NO_MEMORY;
+            return status;
+        }
+
+        accept->listener = listener;
+        accept->ep       = NULL; /* TODO create and wireup an endpoint */
+
+        /* defer user callback to be invoked from the main thread */
+        prog_id = UCS_CALLBACKQ_ID_NULL;
+        uct_worker_progress_register_safe(listener->wiface.worker->uct,
+                                          ucp_listener_conn_request_progress,
+                                          accept, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                          &prog_id);
+    }
+
+    return 0;
+}
+
+ucs_status_t ucp_worker_listen(ucp_worker_h worker,
+                               const ucp_worker_listener_params_t *params,
+                               ucp_listener_h *listener_p)
+{
+    ucp_context_h context = worker->context;
+    ucp_tl_resource_desc_t *resource;
+    uct_iface_params_t iface_params;
+    ucp_listener_h listener = NULL;
+    ucp_rsc_index_t tl_id;
+    ucs_status_t status;
+    ucp_tl_md_t *tl_md;
+    char saddr_str[UCS_SOCKADDR_STRING_LEN];
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&worker->mt_lock);
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    if (!(params->field_mask & UCP_WORKER_LISTENER_PARAM_FIELD_SOCK_ADDR)) {
+        ucs_error("Missing sockaddr for listener");
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
+
+    UCP_CHECK_PARAM_NON_NULL(params->sockaddr.addr, status, goto out);
+
+    /* Go through all the available resources and for each one, check if the given
+     * sockaddr is accessible from its md. Start listening on the first md that
+     * satisfies this.
+     * */
+    for (tl_id = 0; tl_id < context->num_tls; ++tl_id) {
+        resource = &context->tl_rscs[tl_id];
+        tl_md    = &context->tl_mds[resource->md_index];
+
+        if (!(tl_md->attr.cap.flags & UCT_MD_FLAG_SOCKADDR) ||
+            !uct_md_is_sockaddr_accessible(tl_md->md, &params->sockaddr,
+                                           UCT_SOCKADDR_ACC_LOCAL)) {
+            continue;
+        }
+
+        listener = ucs_malloc(sizeof(*listener), "ucp_listener");
+        if (listener == NULL) {
+            status = UCS_ERR_NO_MEMORY;
+            goto out;
+        }
+
+        if (params->field_mask & UCP_WORKER_LISTENER_PARAM_FIELD_CALLBACK) {
+            UCP_CHECK_PARAM_NON_NULL(params->ep_accept_handler.cb, status, goto err_free);
+            listener->cb  = params->ep_accept_handler.cb;
+            listener->arg = params->ep_accept_handler.arg;
+        } else {
+            /* TODO if callback is not specified, avoid creating endpoint on server side
+             * if not really needed (not p2p tl)
+             */
```

Comment:
we are going to create an ep anyway right? for sending a conn_req to the client.

Reply:
right, will remove the comment